reader_concurrency_semaphore.cc

```diff
@@ -1632,7 +1632,13 @@ future<> reader_concurrency_semaphore::with_permit(schema_ptr schema, const char
     auto permit = reader_permit(*this, std::move(schema), std::string_view(op_name), {1, static_cast<ssize_t>(memory)}, timeout, std::move(trace_ptr));
     permit->aux_data().func = std::move(func);
     permit->aux_data().permit_keepalive = permit;
-    return do_wait_admission(*permit);
+    auto fut = futurize_invoke([this, &permit] { return do_wait_admission(*permit); });
+    if (fut.available() && fut.failed()) {
+        // We failed to queue up the permit, it will never get to the execution
+        // loop, release the self-reference to avoid the permit becoming a zombie.
+        auto keepalive = std::exchange(permit->aux_data().permit_keepalive, std::nullopt);
+    }
```

Comment:
Maybe we should address it from the other direction, and avoid the self-reference. Why not give the permit smart-pointer to do_wait_admission and make it responsible for it?

I see that wait_readmission() calls do_wait_admission() without the smart pointer, but perhaps we can have two overloads, one calling the other.

Reply:
In fact wait_readmission is called from the smart pointer one frame back.

Reply:
Actually I think it's just unnecessary. The permit is always held by someone outside the queue, since things like obtain_permit() return it.

The permit handle destructor can destroy the implementation if we want, or the queue can just ignore permits with no attached handle and no func if we want.

Reply:
Not in the case of `with_permit()`, notice how the permit is created in `with_permit()` call and the caller will only get it once `func` is called, which is after admission. The self reference is to keep the permit alive until `func` is called.
Maybe I can do something with `with_permit()` to make the caller hold on to it somehow. I'll see what I can do, I hate this self-reference, it is a landmine we already stepped onto.

Reply:
Dropped the self-reference in v2.